bazooka/tx.go

```diff
@@ -334,3 +340,91 @@ func (b *Bazooka) RegisterPubkeys(pubkeys [16][4]*big.Int) (txHash string, err e
 
 	return tx.Hash().String(), nil
 }
+
+// ApproveToken approves erc20 token
+func (b *Bazooka) ApproveToken(tokenAddr common.Address, spender common.Address, amount uint64) (txHash string, err error) {
+	tokenABI, err := abi.JSON(strings.NewReader(erc20.Erc20ABI))
+	if err != nil {
+		return
+	}
+
+	amountBigInt := big.NewInt(int64(amount))
+	data, err := tokenABI.Pack("approve", spender, amountBigInt)
+	if err != nil {
+		b.log.Error("Error packing data for token approve", "err", err)
+		return
+	}
+
+	tx, err := b.SignAndBroadcast(b.EthClient, tokenAddr, big.NewInt(0), data)
+	if err != nil {
+		b.log.Error("Error sending register batch", "err", err)
+		return
+	}
+
+	b.log.Info("Token approved", "tokenAdrrress", tokenAddr, "txHash", tx.Hash().String())
+
+	return tx.Hash().String(), nil
+
+}
+
+func (b *Bazooka) Deposit(pubkey [4]*big.Int, tokenID uint64, amount uint64) (txHash string, err error) {
+	depositmanagerABI, err := abi.JSON(strings.NewReader(depositmanager.DepositmanagerABI))
+	if err != nil {
+		return
+	}
+	registryABI, err := abi.JSON(strings.NewReader(accountregistry.AccountregistryABI))
+	if err != nil {
+		return
+	}
+	input, err := registryABI.Pack("register", pubkey)
+	if err != nil {
+		b.log.Error("Error packing data for register batch", "err", err)
+		return
+	}
+	fromAddress := b.operator
+	toAddress := ethCmn.HexToAddress(b.Cfg.AccountRegistry)
+	value := big.NewInt(0)
+	callMsg := ethereum.CallMsg{
+		From:  fromAddress,
+		To:    &toAddress,
+		Data:  input,
+		Value: value,
+	}
+	output, err := b.EthClient.CallContract(context.Background(), callMsg, nil)
+	if err != nil {
+		fmt.Println("Error calling contract: ", err)
+		return
+	}
+	outputMap := map[string]interface{}{}
+	err = registryABI.UnpackIntoMap(outputMap, "register", output)
+	if err != nil {
+		fmt.Println("Error unpacking", err)
+		return
+	}
+
+	var pubkeyID *big.Int
+	for _, v := range outputMap {
+		pubkeyID = v.(*big.Int)
+	}
+
+	tx, err := b.SignAndBroadcast(b.EthClient, toAddress, value, input)
```

Comment:
Looks like the purpose here is to send a transaction to register a public key to the account registry. But why we are parsing ABI here? Don't we have AccountregistryTransactor in `accountregistry.go` that we can easily call the `Register` method?

Reply:
Yes we do, but I have this generic transaction sender that I use everywhere, we anyways need to manually pack input data for estimation of gas.

So this is the most generic way I have found to perform transactions, we do lose some really nice type checking which sucks.